File: Phan_B/bai39/bai_39.jl
function cach1(a)
    n = length(a)
    max_sum = -typemax(Int128)
    for i in 1:n
        for j in i:n
            s = sum(a[i:j])
            if s > max_sum; max_sum = s; end
        end
    end
    return max_sum
end

function cach2(a)
    n = length(a)
    max_sum = -typemax(Int128)
    for i in 1:n
        s = 0
        for j in i:n
            s += a[j]
            if s > max_sum; max_sum = s; end
        end
    end
    return max_sum
end

function kadane(a)
    max_sum = a[1]
    current_sum = a[1]
    for i in 2:length(a)
        current_sum = max(a[i], current_sum + a[i])
        max_sum = max(max_sum, current_sum)
    end
    return max_sum
end

input_data = split(read(stdin, String))
if !isempty(input_data)
    n = parse(Int, input_data[1])
    a = [parse(Int, input_data[i]) for i in 2:(n+1)]

    t1 = @elapsed r1 = cach1(a)
    println("Cach 1 time: $(t1)s (Ket qua: $r1)")

    t2 = @elapsed r2 = cach2(a)
    println("Cach 2 time: $(t2)s (Ket qua: $r2)")

    t3 = @elapsed r3 = kadane(a)
    println("Cach 3 time: $(t3)s (Ket qua: $r3)")
end

File: Phan_B/bai39/bai_39.py
import sys
import time

def cach1(a):
    n = len(a)
    max_sum = -float('inf')
    for i in range(n):
        for j in range(i, n):
            s = sum(a[i:j+1])
            if s > max_sum: max_sum = s
    return max_sum

def cach2(a):
    n = len(a)
    max_sum = -float('inf')
    for i in range(n):
        s = 0
        for j in range(i, n):
            s += a[j]
            if s > max_sum: max_sum = s
    return max_sum

def kadane(a):
    max_sum = a[0]
    current_sum = a[0]
    for x in a[1:]:
        current_sum = max(x, current_sum + x)
        max_sum = max(max_sum, current_sum)
    return max_sum

if __name__ == "__main__":
    input_data = sys.stdin.read().split()
    if input_data:
        n = int(input_data[0])
        a = [int(x) for x in input_data[1:n+1]]

        for name, func in [("Cach 1", cach1), ("Cach 2", cach2), ("Kadane", kadane)]:
            start = time.perf_counter()
            res = func(a)
            end = time.perf_counter()
            print(f"{name} time: {end - start:.6f}s (Ket qua: {res})")